src/utils/runWithConcurrentLimit.ts

```diff
@@ -0,0 +1,39 @@
+import * as CliProgress from 'cli-progress';
+
+export async function runWithConcurrentLimit<I>(
+    maxBatchSize: number,
+    inputs: I[],
+    cb: (input: I) => Promise<void>,
+    progress: boolean
+): Promise<void> {
+    const i = [...inputs];
+    let completedJobs = 0;
+    const initialWorkingSet = i.splice(0, maxBatchSize);
+
+    let progressBar: CliProgress.SingleBar | undefined;
+    if (progress) {
+        progressBar = new CliProgress.SingleBar(
+            {
+                etaBuffer: maxBatchSize,
+            },
+            CliProgress.Presets.shades_grey
+        );
+        progressBar.start(inputs.length, 0);
+    }
+
+    const queueNext = (): Promise<void> | void => {
+        const next = i.shift();
+        completedJobs += 1;
+        if (progressBar) {

```

Comment:
You should be able to simplify this with optional chaining `progressBar?.(...)`

(And same below.)

Reply:
My personal preference is to use explicit control flow structures for effects but happy to change this to keep with the project's preferences.